lib/error.sh

```diff
@@ -0,0 +1,218 @@
+		# Hence, delete the first field (PPID) and what
+		# remains is the entire command line.
+		local cmd=$(echo "$details" |\
+			awk '{$1=""; print $0}' |\
+			sed \
+			-e 's/^ *//g' \
+			-e 's/ *$//g')
+
+		msg=$(printf "  %d: {pid: %d, command: '%s'%s}" \
+			"${i}" \
+			"${pid}" \
+			"${cmd}" \
+			"${current}")
+		stderr "$msg"
+
+		[ "$pid" = 1 ] && break
+
+		pid="$ppid"
+	done
+}
+
+show_stacktrace()
+{
+	local err_line="${1:-}"
+	local err_func="${2:-}"
+	local err_path="${3:-}"
+
+	[ -z "$err_line" ] && _fatal "need error location line number"
+	[ -z "$err_func" ] && _fatal "need error location func"
+	[ -z "$err_path" ] && _fatal "need error location file path"
+
+	local line
+	local func
+	local file
+
+	local -i i
+
+	stderr "stacktrace:"
+
+	for ((i = 0; ; i++))
+	do
+		local result
+		result=$(caller "$i" || true)
+
+		[ -z "$result" ] && break
+
+		line=$(echo "$result"|awk '{print $1}')
+		func=$(echo "$result"|awk '{print $2}')
+		file=$(echo "$result"|awk '{print $3}')
+
+		local path
+		path=$(resolve_path "$file")
+
+		local msg
+
+		local current=''
+
+		# Add a visual marker showing where the original error was
+		# detected.
+		[ "${line}" = "${err_line}" ] && \
+		[ "${func}" = "${err_func}" ] && \
+		[ "${path}" = "${err_path}" ] && \
+		current=", current='yes'"
+
+		msg=$(printf "  %d: {function: '%s', file: '%s', line: %d%s}\n" \
+			"${i}" \
+			"${func}" \
+			"${path}" \
+			"${line}" \
+			"${current}" )
+
+		stderr "$msg"
+	done
+}
+
+# Function to be called by die() or a trap/signal handler to dump all
+# details of the environment (in YAML format), to help with debugging.
+dump_details()
+{
+	local err_line="${1:-}"
+	local err_func="${2:-}"
+	local err_path="${3:-}"
+
+	[ -z "$err_line" ] && _fatal "need error location line number"
+	[ -z "$err_func" ] && _fatal "need error location func"
+	[ -z "$err_path" ] && _fatal "need error location file path"
+
+	# Spacer
+	stderr
+
+	stderr "script:"
+	stderr "  name: '$0'"
+	stderr "  pid: $$"
+	stderr "  directory: '$PWD'"
+	stderr "  details: '$(ls -dlZ "$PWD")'"
+	stderr "failure:"
+	stderr "  function: '$func'"
+	stderr "  file: '$path'"
+	stderr "  line: $line"
+	stderr "  name: '$0'"
+
+	show_stacktrace \
+		"${err_line}" \
+		"${err_func}" \
+		"${err_path}"
+
+	show_proc_hierarchy
+
+	stderr "time: '$(date -Isec)'"
+	stderr "runtime-seconds: ${SECONDS}"
+	stderr "host:"
+	stderr "  name: '$(hostname)'"
+	stderr "  uname: '$(uname -a)'"
+
+	stderr "locale:"
+	locale 2>/dev/null | sed \
+		-e 's/^/  /g' \
+		-e 's/=/: '\''/g' \
+		-e 's/$/'\''/g' \
+		>&2
+
+	stderr "user:"
+	stderr "  uid: {value: $UID, name: '$(getent passwd "$UID"|cut -d: -f1)'}"
+	stderr "  euid: {value: $EUID , name: '$(getent passwd "$EUID"|cut -d: -f1)'}"
+	stderr "  groups: '$(id)'"
+
+	stderr "bash:"
+	stderr "  version: '${BASH_VERSION}'"
+	stderr "  version-info: '${BASH_VERSINFO[*]}'"
+
+	stderr "environment:"
+
+	# Remove bash functions (that can span multiple lines)
+	env |\
```

Comment:
😳

I searched longe and hard to find something slightly simpler, and the simplest I can come up with is:

```
env $(env | sed -e "s/\(BASH_FUNC_.*%%\)=().*/-u \1/g;t;d" ) 
```

The way this works is
* Run the original env through a sed that replaces any `BASH_FUNC_foo%%=...` with `-u BASH_FUNC_foo%%`
* Pass these various `-u` options to `env`…
* without a command, so it still prints the environment.